source/common/network/buffered_io_socket_handle_impl.cc

```diff
@@ -252,16 +252,38 @@ Address::InstanceConstSharedPtr BufferedIoSocketHandleImpl::peerAddress() {
   throw EnvoyException(fmt::format("getsockname failed for BufferedIoSocketHandleImpl"));
 }
 
-Event::FileEventPtr BufferedIoSocketHandleImpl::createFileEvent(Event::Dispatcher& dispatcher,
-                                                                Event::FileReadyCb cb,
-                                                                Event::FileTriggerType trigger_type,
-                                                                uint32_t events) {
-  io_callback_ = dispatcher.createSchedulableCallback([this]() { user_file_event_->onEvents(); });
-  auto event = Event::UserSpaceFileEventFactory::createUserSpaceFileEventImpl(
-      dispatcher, cb, trigger_type, events, *io_callback_);
-  user_file_event_ = event.get();
-  return event;
+void BufferedIoSocketHandleImpl::initializeFileEvent(Event::Dispatcher& dispatcher,
+                                                     Event::FileReadyCb cb,
+                                                     Event::FileTriggerType trigger,
+                                                     uint32_t events) {
+  ASSERT(user_file_event_ == nullptr, "Attempting to initialize two `file_event_` for the same "
+                                      "file descriptor. This is not allowed.");
+  ASSERT(trigger == Event::FileTriggerType::Edge, "Only support edge type.");
```

Comment:
This is a bit awkward since windows continues to insist in asking for level events for sockets.

Reply:
Yeah.. The overall windows implementation is not perfect. This could be fixed when another level triggered event implementation is provided.

Can I add a failed on windows test that explicitly announce platform trigger type(level trigger on windows) is not supported yet?

Reply:
@davinci26 

This seems to me like a failure in the way we request the file event. We could instead request something like "EdgeIfSupportedElseLevel".   https://github.com/envoyproxy/envoy/pull/13787 is the start of making level events efficient, but you could also think of it as edge trigger emulation at the envoy proxy layer. My message is that callers that end up passing Level to sockets of this type would be quite happy to get back an edge triggered socket instead.

Reply:
@antoniovicente @davinci26  

It's fairly easy to implement level trigger on the correct edge trigger event system.
I don't want to expand this PR any more. I can make user space event support level trigger when this PR is merged.

Reply:
I am not sure I have the full context here but why does this feature require only edge based events? Is there a discussion on that?

Reply:
This socket handle is not *depending* on edge triggered event :)
This socket handle is *introducing* a new event type and only "edge trigger" parameter is implemented as of today.

Reply:
I see, I get what you are doing on a high level now.

Reply:
See my original comment.  FileEvent backed by libevent need to be registered in level mode when running under windows.  The same is not true for this userspace socket type.  The problem is with the arguments passed in to initializeFileEvent.

Reply:
Nowadays on windows `initializeFileEvent` invokes `EmulatedEdge`, not `Level`.
The emulation is a FileEventImpl and IoSocketHandleImpl internal behavior.
From the perspective of UserspaceEvent and UserspaceIoHandle emulatedEdge is a no-op. A edge-triggered event don't need any extra operation to emulate edge triggered behavior :)

Indeed I want to support the native level triggered user space event but I doubt if it is needed...